### PR TITLE
*: fix BatchPointGet return empty value for already locked key.

### PR DESCRIPTION
**Pull request by coocood**
### What problem does this PR solve? 
Fixes https://github.com/pingcap/tidb/issues/15361

### What is changed and how it works?
- Do not use the returned values from the pessimistic lock directly.
- Put the returned value in the TxnCtx cache and put the cache into the batch getter as a middle cache.

### Check List 

Tests 

 - Unit test

Code changes

 - Has interface methods change

Release note

 - fix BatchPointGet return empty value for an already locked key.

### CODE REVIEW

File: executor/point_get.go [hunk 1]
```diff
@@ -181,15 +181,17 @@ func (e *PointGetExecutor) lockKeyIfNeeded(ctx context.Context, key []byte) erro
...
+		lockCtx.Values = map[string]kv.ReturnedValue{}
 		err := doLockKeys(ctx, e.ctx, lockCtx, key)
 		if err != nil {
 			return err
 		}
 		lockCtx.ValuesLock.Lock()
 		defer lockCtx.ValuesLock.Unlock()
 		for key, val := range lockCtx.Values {
-			seVars.TxnCtx.SetPessimisticLockCache(kv.Key(key), val)
+			if !val.AlreadyLocked {
```
Comment (cfzjywxk) [inline executor/point_get.go]: Will this lead to reading incorrect data?

begin

point get lock key1,  the lock operation succeeded, but the following operations of this statement failed  for some reasons, the value is cached to txn middle cache, but the statement is async rollbacked

batch point get lock again,  reading values from the middle cache, but the value is invalid

Or shoud we just let pessimistic lock always return value in the pessimistic lock response

Comment (coocood) [reply executor/point_get.go]: The pessimistic lock response does always return value, but TiDB would send lock request if it's already locked.

If the key is successfully locked, then we will never do PessimisticRollback on this key even if there are other errors for the statement.

**Comment by cfzjywxk** [comment]
LGTM

File: session/pessimistic_test.go [hunk 2]
```diff
@@ -1097,6 +1097,17 @@ func (s *testPessimisticSuite) TestBatchPointGetLockIndex(c *C) {
...
 }
 
+func (s *testPessimisticSuite) TestBatchPointGetAlreadyLocked(c *C) {
+	tk := testkit.NewTestKitWithInit(c, s.store)
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t (c1 int, c2 int, c3 int, primary key(c1, c2))")
+	tk.MustExec("insert t values (1, 1, 1), (2, 2, 2)")
+	tk.MustExec("begin pessimistic")
+	tk.MustExec("select * from t where c1 > 1 for update")
+	tk.MustExec("select * from t where (c1, c2) in ((2,2)) for update")
```
Comment (tiancaiamao) [inline session/pessimistic_test.go]: CheckRow?

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**